lib/main/atom/views/semanticView.tsx

```diff
@@ -0,0 +1,424 @@
+
+/**
+ * Eventually
+ */
+namespace rts {
+  /** 0 based length */
+  export function indent(indent: number) {
+    return Array(indent + 1).join().split('').map(i => "\u00a0\u00a0\u00a0\u00a0");
+  }
+}
+
+//TODO extract?
+interface Position {
+  line: number;
+  offset: number;
+}
+
+interface SemanticTreeNode {
+  text: string;
+  kind: string;
+  kindModifiers: string;
+  spans: Array<{start: Position, end: Position}>;
+  childItems?: Array<SemanticTreeNode>;
+}
+
+export interface TypescriptFileBuffer {
+  buffer: TypescriptBuffer;
+  isOpen: boolean;
+}
+
+class SemanticViewRenderer {
+
+  editor: AtomCore.IEditor;
+  navTree: SemanticTreeNode | null;
+  getTypescriptBuffer: (filePath: string) => Promise<TypescriptFileBuffer>;
+  root: HTMLElement;
+  selectedNode: HTMLElement | null;
+
+  constructor(options: {getTypescriptBuffer: (filePath: string) => Promise<TypescriptFileBuffer>}) {
+    this.getTypescriptBuffer = options.getTypescriptBuffer;
+    this.navTree = null;
+    this.selectedNode = null;
+  }
+
+  setEditor(editor: AtomCore.IEditor) {
+    this.editor = editor;
+  }
+
+  setNavTree(navTree: SemanticTreeNode | null) {
+    this.navTree = navTree;
+    this._render();
+  }
+
+  forceUpdate() {
+    //FIXME should this be done differently?
+    this._render();
+  }
+
+  loadNavTree(buffer: TypescriptFileBuffer, filePath?: string) {
+    filePath = filePath ? filePath : this.editor.getPath();
+    buffer.buffer.getClient(filePath).then(client => client.executeNavTree({file: filePath as string}).then(navtreeResult => {
+      let navTree = navtreeResult ? navtreeResult.body as SemanticTreeNode : void (0);
+      if (navTree) {
+        this.setNavTree(navTree);
+      }
+    }).catch(err => console.error(err, filePath))
+    );
+  }
+
+  componentDidMount() {
+    // We listen to a few things
+
+    // Editor scrolling
+    var editorScrolling: AtomCore.Disposable;
+    // Editor changing
+    var editorChanging: AtomCore.Disposable;
+
+    let subscribeToEditor = (editor: AtomCore.IEditor) => {
+      this.setEditor(editor);
+
+      //set navTree
+      const filePath = editor.getPath();
+      this.getTypescriptBuffer(filePath).then(fbuffer => {
+        let buffer = fbuffer as TypescriptFileBuffer;
+        if (buffer.isOpen) {
+          this.loadNavTree(buffer, filePath);
+        } else {
+          buffer.buffer.open().then(() => {
+            this.loadNavTree(buffer, filePath);
+          }).catch(err => console.error(err, filePath));
+        }
+      });
+      // Subscribe to stop scrolling
+      editorScrolling = editor.onDidChangeCursorPosition(() => {
+        this.selectAtCursorLine();
+      });
+
+
+      editorChanging = editor.onDidStopChanging(() => {
+        //set navTree
+        const filePath = editor.getPath();
+        this.getTypescriptBuffer(filePath).then(buffer => {
+          if (buffer.isOpen) {
+            this.loadNavTree(buffer, filePath);
```

Comment:
To answer your question, you only need to use the TypescriptBuffer wrapper if you plan to make changes to the source file. This is so `tsserver` is kept in the loop of any changes and can update its internal representation. In this case you don't even need to know the content of the file so it's not necessary.